src/app/clusters/service-area-server/service-area-delegate.h

```diff
@@ -0,0 +1,389 @@
+     *
+     * The caller of this method will ensure that all the locations in the set are valid supported locations.
+     *
+     * @param[in] req List of new selected locations.
+     * @param[out] locationStatus Success if all checks pass, error code if failure.
+     * @param[out] statusText text describing failure (see description above), size kMaxSizeStatusText + 1 byte for terminating
+     * character.
+     * @param[out] useStatusText if true, the statusText value should be returned in the command response.
+     * @return true if success.
+     *
+     * @note If the SelectLocations command is allowed when the device is operating and the selected locations change to none, the
+     * device must stop.
+     */
+    virtual bool IsValidSelectLocationsSet(const Commands::SelectLocations::DecodableType & req,
+                                           SelectLocationsStatus & locationStatus, char * statusText, bool & useStatusText) = 0;
+
+    /**
+     * @brief The server instance ensures that the SelectedLocations and CurrentLocation attributes are not null before
+     * calling this method.
+     * @param[out] skipStatusText text describing why current location cannot be skipped.
+     * @return true if command is successful, false if the received skip request cannot be handled due to the current mode of the
+     * device.
+     *
+     * @note skipStatusText must be filled out by the function on failure.
+     *
+     * @note If the device successfully accepts the request and the ListOrder feature is set to 1:
+     *       The server SHALL stop operating at the current location.
+     *       The server SHALL attempt to operate at the remaining locations on the SelectedLocations attribute list, starting with
+     * the next entry. If the end of the SelectedLocations attribute list is reached, the server SHALL stop operating.
+     *
+     * @note If the device successfully accepts the request and the ListOrder feature is set to 0:
+     *       The server SHALL stop operating at the current location.
+     *       The server SHALL attempt to operate at the locations on the SelectedLocations attribute list where operating has not
+     * been completed, using a vendor defined order. If the server has completed operating at all locations on the SelectedLocations
+     * attribute list, the server SHALL stop operating.
+     *
+     * @note If the Status field is set to InvalidLocationList, the StatusText field SHALL be an empty string.
+     *       If the Status field is not set to Success, or InvalidLocationList, the StatusText field SHALL include a vendor defined
+     * error description which can be used to explain the error to the user. For example, if the Status field is set to
+     * InvalidInMode, the StatusText field SHOULD indicate why the request is not allowed, given the current mode of the device,
+     * which may involve other clusters.
+     */
+    virtual bool HandleSkipCurrentLocation(char * skipStatusText)
```

Comment:
I am stopping here.  All the `char*` needs to go away, because that's a good way to get security bugs.

Reply:
This has been addressed in [this commit](https://github.com/project-chip/connectedhomeip/pull/33991/commits/38a7e7a9c206391ede4528ffd1bc73c391c400a1).